File: app/src/main/java/org/sil/storyproducer/tools/media/story/StoryFrameDrawer.java
package org.sil.storyproducer.tools.media.story;

import android.graphics.Bitmap;
import android.graphics.Canvas;
import android.graphics.Paint;
import android.graphics.Rect;
import android.media.MediaFormat;
import android.util.Log;

import org.sil.storyproducer.tools.media.MediaHelper;
import org.sil.storyproducer.tools.media.pipe.PipedVideoSurfaceEncoder;
import org.sil.storyproducer.tools.media.pipe.SourceUnacceptableException;

import java.io.IOException;

/**
 * This class knows how to draw the frames provided to it by {@link StoryMaker}.
 */
class StoryFrameDrawer implements PipedVideoSurfaceEncoder.Source {
    private static final String TAG = "StoryFrameDrawer";

    private final MediaFormat mVideoFormat;
    private final StoryPage[] mPages;
    private final long mAudioTransitionUs;
    private final long mSlideTransitionUs;

    private final int mFrameRate;

    private final int mWidth;
    private final int mHeight;
    private final Rect mScreenRect;

    private int mCurrentSlideIndex = -1; //starts at -1 to allow initial transition
    private long mCurrentSlideExDuration = 0; //exclusive duration of current slide
    private long mCurrentSlideStart = 0; //time (after transition) of audio start

    private long mCurrentSlideImgDuration = 0; //total duration of current slide image (cached for performance)
    private long mNextSlideImgDuration = 0; //total duration of next slide image (cached for performance)

    private int mCurrentFrame = 0;

    private boolean mIsVideoDone = false;

    StoryFrameDrawer(MediaFormat videoFormat, StoryPage[] pages, long audioTransitionUs, long slideTransitionUs) {
        mVideoFormat = videoFormat;
        mPages = pages;

        mAudioTransitionUs = audioTransitionUs;

        long correctedSlideTransitionUs = slideTransitionUs;

        //mSlideTransition must never exceed the length of slides in terms of audio.
        //Pre-process pages and clip the slide transition time to fit in all cases.
        for(StoryPage page : pages) {
            long totalPageUs = page.getAudioDuration() + mAudioTransitionUs;
            if(correctedSlideTransitionUs > totalPageUs) {
                correctedSlideTransitionUs = totalPageUs;
                Log.d(TAG, "Corrected slide transition from " + slideTransitionUs + " to " + correctedSlideTransitionUs);
            }
        }

        mSlideTransitionUs = correctedSlideTransitionUs;

        mFrameRate = mVideoFormat.getInteger(MediaFormat.KEY_FRAME_RATE);

        mWidth = mVideoFormat.getInteger(MediaFormat.KEY_WIDTH);
        mHeight = mVideoFormat.getInteger(MediaFormat.KEY_HEIGHT);
        mScreenRect = new Rect(0, 0, mWidth, mHeight);
    }

    private void drawFrame(Canvas canv, int pageIndex, long timeOffsetUs, long imgDurationUs, float alpha) {
        //In edge cases, draw a black frame with alpha value.
        if(pageIndex < 0 || pageIndex >= mPages.length) {
            canv.drawARGB((int) (alpha * 255), 0, 0, 0);
            return;
        }

        StoryPage page = mPages[pageIndex];
        Bitmap bitmap = page.getBitmap();
        KenBurnsEffect kbfx = page.getKenBurnsEffect();

        float position = (float) (timeOffsetUs / (double) imgDurationUs);
        Rect drawRect = kbfx.interpolate(position);

        if (MediaHelper.VERBOSE) {
            Log.v(TAG, "drawer: drawing rectangle (" + drawRect.left + ", " + drawRect.top + ", "
                    + drawRect.right + ", " + drawRect.bottom + ") of bitmap ("
                    + bitmap.getWidth() + ", " + bitmap.getHeight() + ")");
        }

        Paint p = new Paint(0);
        p.setAntiAlias(true);
        p.setFilterBitmap(true);
        p.setDither(true);
        p.setAlpha((int) (alpha * 255));

        canv.drawBitmap(bitmap, drawRect, mScreenRect, p);
    }

    @Override
    public long fillCanvas(Canvas canv) {
        long currentTimeUs = MediaHelper.getTimeFromIndex(mFrameRate, mCurrentFrame);

        long nextSlideTransitionUs = mSlideTransitionUs;
        //For pre-first "slide" and last slide, make the transition half as long.
        if(mCurrentSlideIndex == -1 || mCurrentSlideIndex == mPages.length - 1) {
            nextSlideTransitionUs /= 2;
        }

        long nextSlideTransitionStartUs = mCurrentSlideStart + mCurrentSlideExDuration;

        while(currentTimeUs > nextSlideTransitionStartUs + nextSlideTransitionUs) {
            mCurrentSlideIndex++;

            if(mCurrentSlideIndex >= mPages.length) {
                mIsVideoDone = true;
                break;
            }

            mCurrentSlideStart = mCurrentSlideStart + mCurrentSlideExDuration + nextSlideTransitionUs;
            mCurrentSlideExDuration = mPages[mCurrentSlideIndex].getAudioDuration() + mAudioTransitionUs - mSlideTransitionUs;
            mCurrentSlideImgDuration = mNextSlideImgDuration;
            nextSlideTransitionStartUs = mCurrentSlideStart + mCurrentSlideExDuration;

            if(mCurrentSlideIndex + 1 < mPages.length) {
                mNextSlideImgDuration = mPages[mCurrentSlideIndex + 1].getAudioDuration() + 2 * mSlideTransitionUs;
            }
        }

        long timeSinceCurrentSlideStartUs = currentTimeUs - mCurrentSlideStart;
        long currentSlideOffsetUs = timeSinceCurrentSlideStartUs + mSlideTransitionUs;

        drawFrame(canv, mCurrentSlideIndex, currentSlideOffsetUs, mCurrentSlideImgDuration, 1);

        if(currentTimeUs > nextSlideTransitionStartUs) {
            long timeSinceTransitionStartUs = currentTimeUs - nextSlideTransitionStartUs;
            long extraOffsetUs = mSlideTransitionUs - nextSlideTransitionUs; //0 normally, transition/2 for edge cases
            long nextOffsetUs = timeSinceTransitionStartUs + extraOffsetUs;
            drawFrame(canv, mCurrentSlideIndex + 1, nextOffsetUs, mNextSlideImgDuration, nextOffsetUs / (float) nextSlideTransitionUs);
        }

        mCurrentFrame++;

        return currentTimeUs;
    }

    @Override
    public MediaHelper.MediaType getMediaType() {
        return MediaHelper.MediaType.VIDEO;
    }

    @Override
    public void setup() throws IOException, SourceUnacceptableException {
        if(mPages.length > 0) {
            mNextSlideImgDuration = mPages[0].getAudioDuration() + 2 * mSlideTransitionUs;
        }
    }

    @Override
    public MediaFormat getOutputFormat() {
        return mVideoFormat;
    }

    @Override
    public boolean isDone() {
        return mIsVideoDone;
    }

    @Override
    public void close() {
        //Do nothing.
    }
}

File: app/src/main/java/org/sil/storyproducer/tools/media/story/StoryFrameDrawer.kt
package org.sil.storyproducer.tools.media.story

import android.content.Context
import android.graphics.*
import android.media.MediaFormat
import android.text.Layout
import android.util.Log
import org.sil.storyproducer.tools.file.getStoryImage

import org.sil.storyproducer.tools.media.MediaHelper
import org.sil.storyproducer.tools.media.graphics.TextOverlay
import org.sil.storyproducer.tools.media.pipe.PipedVideoSurfaceEncoder
import org.sil.storyproducer.tools.media.pipe.SourceUnacceptableException

import java.io.IOException

/**
 * This class knows how to draw the frames provided to it by [StoryMaker].
 */
internal class StoryFrameDrawer(private val context: Context, private val mVideoFormat: MediaFormat, private val mPages: Array<StoryPage>, private val mAudioTransitionUs: Long, slideCrossFadeUs: Long) : PipedVideoSurfaceEncoder.Source {
    private val xTime: Long //transition (cross fade) time

    private val mFrameRate: Int

    private val mWidth: Int
    private val mHeight: Int

    private val mBitmapPaint: Paint

    private var mCurrentTextOverlay: TextOverlay? = null
    private var mNextTextOverlay: TextOverlay? = null

    private var slideIndex = -1 //starts at -1 to allow initial transition
    private var slideAudioStart: Long = 0
    private var slideAudioEnd: Long = 0
    private var nSlideAudioEnd: Long = 0
    private val slideVisStart: Long
        get() {return if(slideIndex<=0){slideAudioStart} else {slideAudioStart-xTime/2}}
    private val slideXStart: Long  //beginning of the next transition
        get() {return if(slideIndex>=mPages.size-1){slideAudioEnd} else {slideAudioEnd-xTime/2}}
    private val slideXEnd: Long  //end of the next transition
        get() {return if(slideIndex>=mPages.size-1){slideAudioEnd} else {slideAudioEnd+xTime/2}}
    private val nSlideXEnd: Long  //end of the next transition
        get() {return if(slideIndex>=mPages.size-2){nSlideAudioEnd} else {nSlideAudioEnd+xTime/2}}
    private val slideVisDur: Long // the visible duration of the slide
        get() {return slideXEnd - slideVisStart}
    private val nSlideVisDur: Long // the visible duration of the next slide
        get() {return nSlideXEnd - slideXStart}

    private var mCurrentFrame = 0

    private var mIsVideoDone = false

    private var bitmaps: MutableMap<String,Bitmap?> = mutableMapOf()

    init {

        var correctedSlideTransitionUs = slideCrossFadeUs

        //mSlideTransition must never exceed the length of slides in terms of audio.
        //Pre-process pages and clip the slide transition time to fit in all cases.
        for (page in mPages) {
            val totalPageUs = page.audioDuration + mAudioTransitionUs
            if (correctedSlideTransitionUs > totalPageUs) {
                correctedSlideTransitionUs = totalPageUs
                Log.d(TAG, "Corrected slide transition from $slideCrossFadeUs to $correctedSlideTransitionUs")
            }
        }

        xTime = correctedSlideTransitionUs

        mFrameRate = mVideoFormat.getInteger(MediaFormat.KEY_FRAME_RATE)

        mWidth = mVideoFormat.getInteger(MediaFormat.KEY_WIDTH)
        mHeight = mVideoFormat.getInteger(MediaFormat.KEY_HEIGHT)

        mBitmapPaint = Paint()
        mBitmapPaint.isAntiAlias = true
        mBitmapPaint.isFilterBitmap = true
        mBitmapPaint.isDither = true
    }

    override fun getMediaType(): MediaHelper.MediaType {
        return MediaHelper.MediaType.VIDEO
    }

    override fun getOutputFormat(): MediaFormat {
        return mVideoFormat
    }

    override fun isDone(): Boolean {
        return mIsVideoDone
    }

    @Throws(IOException::class, SourceUnacceptableException::class)
    override fun setup() {
        if (mPages.isNotEmpty()) {
            val nextPage = mPages[0]

            val nextText = nextPage.text
            mNextTextOverlay = TextOverlay(nextText)
            //Push text to bottom if there is a picture.
            mNextTextOverlay!!.setVerticalAlign(Layout.Alignment.ALIGN_OPPOSITE)
        }
    }

    override fun fillCanvas(canv: Canvas): Long {

        //[-|-page-1-|-| ]
        //           [ |-|-page-2-|-| ]
        //                        [ |-|-page-last-|-]
        // | | | (two bars) = transition time (xtime)
        // | | (one bar) = 1/2 xtime
        // --- (dash) sound playing from slide
        // Exclusive time + xtime/2 for first and last slide
        // "current page" is the page until it ends
        // "Next page" is growing in intensity for "xtime"
        // Visible time

        //Each time this is called, go forward 1/30 of a second.
        val cTime = MediaHelper.getTimeFromIndex(mFrameRate.toLong(), mCurrentFrame)

        if(cTime > slideXEnd){
            //go to the next slide
            slideIndex++

            if (slideIndex >= mPages.size) {
                mIsVideoDone = true
            } else {
                slideAudioStart = slideAudioEnd
                slideAudioEnd += mPages[slideIndex].getDuration(mAudioTransitionUs)
                mCurrentTextOverlay = mNextTextOverlay

                if (slideIndex + 1 < mPages.size) {
                    nSlideAudioEnd = slideAudioEnd + mPages[slideIndex + 1].getDuration(mAudioTransitionUs)

                    val nextText = mPages[slideIndex + 1].text
                    mNextTextOverlay = TextOverlay(nextText)
                    //Push text to bottom if there is a picture.
                    mNextTextOverlay!!.setVerticalAlign(Layout.Alignment.ALIGN_OPPOSITE)
                }
            }
        }

        drawFrame(canv, slideIndex, cTime - slideVisStart, slideVisDur,
                1f, mCurrentTextOverlay)

        if (cTime >= slideXStart) {
            val alpha = (cTime - slideXStart) / xTime.toFloat()
            drawFrame(canv, slideIndex + 1, cTime - slideXStart, nSlideVisDur,
                    alpha, mNextTextOverlay)
        }

        //clear image cache to save memory.
        if(slideIndex >= 1) {
            if (bitmaps.containsKey(mPages[slideIndex - 1].imRelPath)) {
                bitmaps.remove(mPages[slideIndex - 1].imRelPath)
            }
        }

        mCurrentFrame++

        return cTime
    }

    private fun drawFrame(canv: Canvas, pageIndex: Int, timeOffsetUs: Long, imgDurationUs: Long,
                          alpha: Float, overlay: TextOverlay?) {
        //In edge cases, draw a black frame with alpha value.
        if (pageIndex < 0 || pageIndex >= mPages.size) {
            canv.drawARGB((alpha * 255).toInt(), 0, 0, 0)
            return
        }

        val page = mPages[pageIndex]
        if(!bitmaps.containsKey(page.imRelPath)){
            bitmaps[page.imRelPath] = getStoryImage(context,page.imRelPath)
        }
        val bitmap = bitmaps[page.imRelPath]

        if (bitmap != null) {
            val kbfx = page.kenBurnsEffect

            val position = (timeOffsetUs / imgDurationUs.toDouble()).toFloat()

            val drawRect: RectF
            if (kbfx != null) {
                drawRect = kbfx.revInterpolate(position,mWidth,mHeight,bitmap.width,bitmap.height)
            } else {
                drawRect = RectF(0f, 0f, mWidth*1f, mHeight*1f)
            }

            mBitmapPaint.alpha = (alpha * 255).toInt()

            canv.drawBitmap(bitmap, null, drawRect, mBitmapPaint)
        } else {
            //If there is no picture, draw black background for text overlay.
            canv.drawARGB((alpha * 255).toInt(), 0, 0, 0)
        }

        if (overlay != null) {
            overlay.setAlpha(alpha)
            overlay.draw(canv)
        }
    }

    override fun close() {
        bitmaps.clear()
    }

    companion object {
        private val TAG = "StoryFrameDrawer"
    }
}
